### DISCUSSION TITLE
IDs for ValueObject types

### DISCUSSION THREAD

**Discussion by cornwe19** [javers/javers#1084] [Q&A] [answered]
I have a situation similar to the following

```kotlin
@Entity
data class QuestionSet(
    @Id
    val id: Int
    @OneToMany
    var links: MutableList<QuestionLink>
)

@ValueObject
@Entity
data class QuestionLink(
    @Id
    val id: Int,
    var order: Int,
    var isRelevantForChildren: Boolean,
    var isRelevantForAdults: Boolean,
    @ShallowReference
    val question: Question
)
```

Within a `QuestionSet`, `QuestionLink`s are both reorderable and mutable. They represent the configuration of the `QuestionSet`, so I'd like edits to them to be tracked as edits to the set itself. The oddity I'm running into is that Javers appears to assign value object IDs based on each `QuestionLink`'s index in the list. You can imagine how the diff gets pretty confusing when questions get reordered during a commit since it'll appear like a particular index changed all of its properties instead of simply reporting that its display order was altered.

Is there a way to declare custom embedded object IDs on `ValueObject`s  which reside in Lists or Sets within their owning entity? Or is there some configuration I could apply to make differencing reordered `ValueObject` lists easier to follow?

**Comment by cornwe19** [comment, marked as answer]
For anyone who stumbles on this in the future, I was able to mostly resolve the issue by treating the ordered list of `QuestionLink`s as a mutable map instead of a mutable list. The new declaration looks something more like this.

```Kotlin
@Entity
data class QuestionSet(
    @Id
    val id: Int
    @OneToMany
    @MapKey(name = "id")
    var links: MutableMap<Int,QuestionLink>
)
```

Javers appears to use the map's key as a local ID for the owned value objects which resolves the identity differencing problems I was having as items' `order` changed within the list.

**Comment by bartoszwalacik** [comment]
Try ListCompareAlgorithm.AS_SET
